Make the Bringer better testable

matejak opened the pull request:
#### Description:

This PR moves functionality formerly located in the data_fetch module to the Bringer, so one can use polymorphism to create great mocks by just subclassing.
Now, it is possible to test handling of threads and fetch methods separately, because Bringer just calls a method to fetch a URI, and then wraps the process in a thread.

#### Rationale:

- The PR decomposes and concentrates the existing functionality, making it easier to test.

#### Review Hints:

- Business as usual, tests pass, bringer still able to copy things, touched functions were not used by other parts of the addon.

matejak committed:
Restructured part of the fetching code

- Extract wait for network to a separate utility function
- Delegate wrapping of a fetch by thread to a method of Bringer
- Remove the now-empty intermediate fetching helper functions from data_fetch

matejak committed:
Add a test for concurrent fetching prevention

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/OpenSCAP/oscap-anaconda-addon/inspections/4cf47354-a673-437f-bbab-0fe26e547714) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/oscap-anaconda-addon/inspections/4cf47354-a673-437f-bbab-0fe26e547714) completed: **5 updated code elements**

jan-cerny commented on org_fedora_oscap/content_discovery.py:
```diff
@@ -112,23 +113,31 @@ def _fetch_files(self, ca_certs_path):
+        fetch_data_thread = AnacondaThread(
+            name=fetching_thread_name,
+            target=self.fetch_operation,
+            args=(self.content_uri, self.dest_file_name, ca_certs_path),
```
I have noticed a weird thing that was there before but I think we can talk about it nevertheless. The URI is a property of the class (`self.content_uri`) but the certificate path isn't a property of the class but a variable (`ca_certs_path`) and is passed down. But, both are arguments of the `fetch_content` content method. So the arguments of `fetch_content` are each handled differently at this moment.

matejak replied:
Good point, the common argument to a generl fetch operation would be the URI that we store in the class anyway, and then other arguments depend on how we fetch, certs being needed only if we fetch from network using an "s" protocol.
Therefore, it is cleaner to store those variable arguments in the instance as well.

jan-cerny commented on org_fedora_oscap/content_discovery.py:
```diff
@@ -112,23 +113,31 @@ def _fetch_files(self, ca_certs_path):
 
-    def finish_content_fetch(self, fetching_thread_name, fingerprint):
+    def fetch_operation(self, uri, out_file, ca_certs_path=None):
+        return data_fetch.fetch_data(uri, out_file, ca_certs_path)
```
What is the reason for having this wrapper function? Can we simply move code from the `data_fetch.fetch_data` to `fetch_operation`?

matejak replied:
The other function is somewhat coupled with the `data_fetch` module - the way it is unit-tested as well as the way exceptions it can raise are defined.
The whole data_fetch module could be merged into the bringer in the future, but I think that it is OK to have such thin wrapper.
The fetch operation as a separate method is convenient to be able to override it in child classes e.g. for testing purposes.

jan-cerny reviewed (commented):
So far it works great for me because I tried to build the update image and run the Anaconda installation and changed various SCAP content in the Security Policy Scope.

However, I have found an issue, which isn't caused by the contents of this PR: https://github.com/OpenSCAP/oscap-anaconda-addon/issues/244

matejak committed:
Make bringer methods use class attributes

- Make ca_certs a class attribute
- Don't pass those items as method arguments all the time,
  simply reference them from the instance.

matejak committed:
Remove a useless variable

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/oscap-anaconda-addon/inspections/1d9334fa-d320-4d19-9ce4-f41973218458) completed: **5 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/oscap-anaconda-addon/inspections/8aeb4a52-68bc-479f-a6bc-a20dcc96f81c) completed: **5 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/OpenSCAP/oscap-anaconda-addon/inspections/684ccdd6-ee93-4d07-8dfe-0c1c161b4ce7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/oscap-anaconda-addon/inspections/684ccdd6-ee93-4d07-8dfe-0c1c161b4ce7) completed: **5 updated code elements**

jan-cerny reviewed (approved):
I have tried to download various content usign the GUI and I have performed a RHEL9.1 installation from GUI and also an unattended installation with a kickstart.
